modules/apps/batch-planner/batch-planner-web/src/main/resources/META-INF/resources/js/TemplateSelect.js

```diff
@@ -0,0 +1,140 @@
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayForm, {ClaySelect} from '@clayui/form';
+import {fetch, openToast} from 'frontend-js-web';
+import PropTypes from 'prop-types';
+import React, {useEffect, useState} from 'react';
+
+import {
+	HEADERS_BATCH_PLANNER_URL,
+	HEADLESS_ENDPOINT_POLICY_NAME,
+	NULL_TEMPLATE_VALUE,
+	TEMPLATE_SELECTED_EVENT,
+	TEMPLATE_SOILED,
+} from './constants';
+
+const TemplateSelect = ({
+	initialTemplate,
+	portletNamespace,
+	templatesOptions,
+}) => {
+	const [selectedTemplateId, setTemplate] = useState(
+		templatesOptions.find((option) => option.selected)?.value
+	);
+
+	useEffect(() => {
+		if (initialTemplate) {
+			Liferay.fire(TEMPLATE_SELECTED_EVENT, {
+				template: initialTemplate,
+			});
+		}
+	}, [initialTemplate]);
+
+	useEffect(() => {
+		function handleTemplateDirty() {
+			setTemplate(NULL_TEMPLATE_VALUE);
+		}
+
+		Liferay.on(TEMPLATE_SOILED, handleTemplateDirty);
+
+		return () => {
+			Liferay.detach(TEMPLATE_SOILED, handleTemplateDirty);
+		};
+	}, []);
+
+	const onChange = (event) => {
+		const newTemplateId = event.target.value;
+		setTemplate(newTemplateId);
+		fireTemplateSelectionEvent(newTemplateId);
+	};
+
+	const selectId = `${portletNamespace}templateName`;
+
+	return (
+		<ClayForm.Group className="form-group-sm">
+			<label htmlFor={selectId}>{Liferay.Language.get('template')}</label>
+
+			<ClaySelect
+				id={selectId}
+				name={selectId}
+				onChange={onChange}
+				value={selectedTemplateId}
+			>
+				<ClaySelect.Option key={0} value={NULL_TEMPLATE_VALUE} />
+
+				{templatesOptions.map((option) => (
+					<ClaySelect.Option
+						key={option.value}
+						label={option.label}
+						value={option.value}
+					/>
+				))}
+			</ClaySelect>
+		</ClayForm.Group>
+	);
+};
+
+TemplateSelect.propTypes = {
+	portletNamespace: PropTypes.string.isRequired,
+	selectedTemplateClassName: PropTypes.string,
+	selectedTemplateHeadlessEndpoint: PropTypes.string,
+	selectedTemplateMapping: PropTypes.object,
+	templatesOptions: PropTypes.arrayOf(PropTypes.object),
+};
+
+function getMappingFromTemplate(template) {
+	const {mappings} = template;
+
+	return mappings.reduce((accumulator, map) => {
+		accumulator[map.internalFieldName] = map.externalFieldName;
+
+		return accumulator;
+	}, {});
+}
+
+async function fireTemplateSelectionEvent(templateId) {
+	if (templateId === NULL_TEMPLATE_VALUE) {
+		return Liferay.fire(TEMPLATE_SELECTED_EVENT, {
+			template: null,
+		});
+	}
+
+	const request = await fetch(
+		`${HEADERS_BATCH_PLANNER_URL}/plans/${templateId}`
+	);
+
+	if (!request.ok) {
```

Comment:
fetch can also throw an error so `request.ok` is good but not enough. In the worst case, you won't manage to get to line 118. Handle it also via `try catch` please